pkg/controller/namespace/namespace_controller_utils.go

```diff
@@ -103,99 +130,145 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
+	for _, groupVersionResource := range groupVersionResources {
+		glog.V(4).Infof("namespace controller - deleteAllContent - deleting resource - namespace: %s, resource: %v", namespace, groupVersionResource)
+		// estimate how long it will take for the resource to be deleted (needed for objects that support graceful delete)
+		groupVersionResourceEstimate, supportsGrace, err := estimateGracefulTermination(kubeClient, groupVersionResource, namespace, namespaceDeletedAt)
 		if err != nil {
+			glog.V(4).Infof("namespace controller - deleteAllContent - unable to estimate - namespace: %s, resource: %v, err: %v", namespace, groupVersionResource, err)
 			return estimate, err
 		}
-		if containsResource(resources, "horizontalpodautoscalers") {
-			err = deleteHorizontalPodAutoscalers(kubeClient.Extensions(), namespace)
-			if err != nil {
-				return estimate, err
-			}
+		if groupVersionResourceEstimate > estimate {
+			estimate = groupVersionResourceEstimate
 		}
-		if containsResource(resources, "ingresses") {
-			err = deleteIngress(kubeClient.Extensions(), namespace)
-			if err != nil {
-				return estimate, err
-			}
+		var deleteOptions *v1.DeleteOptions
+		if supportsGrace {
+			deleteOptions = &v1.DeleteOptions{GracePeriodSeconds: &groupVersionResourceEstimate}
 		}
-		if containsResource(resources, "daemonsets") {
-			err = deleteDaemonSets(kubeClient.Extensions(), namespace)
-			if err != nil {
-				return estimate, err
-			}
+
+		glog.V(4).Infof("namespace controller - deleteAllContent - estimate - namespace: %s, resource: %v, estimate: %v", namespace, groupVersionResource, groupVersionResourceEstimate)
+
+		// get a client for this group version...
+		dynamicClient, err := clientPool.ClientForGroupVersion(groupVersionResource.GroupVersion())
+		if err != nil {
+			glog.V(4).Infof("namespace controller - deleteAllContent - unable to get client - namespace: %s, resource: %v, err: %v", namespace, groupVersionResource, err)
+			return estimate, err
+		}
+
+		// the discovery API does not tell you if a top-level resource supports LIST or DELETE_COLLECTION
+		// rather than consulting swagger, we instead test if a resource supports LIST or DELETE_COLLECTION.
+		// if the resource does not support LIST OR DELETE_COLLECTION, we cache it to avoid making unnecessary calls in future.
+		apiResource := unversioned.APIResource{Name: groupVersionResource.Resource, Namespaced: true}
+		canDeleteCollectionKey := operationKey{
+			op:  operationDeleteCollection,
+			gvr: groupVersionResource,
 		}
-		if containsResource(resources, "jobs") {
-			err = deleteJobs(kubeClient.Extensions(), namespace)
-			if err != nil {
-				return estimate, err
+		canListKey := operationKey{
+			op:  operationList,
+			gvr: groupVersionResource,
+		}
+		canDeleteCollection, canDeleteCollectionKnown := operationSupportedCache[canDeleteCollectionKey]
+		canList, canListKnown := operationSupportedCache[canListKey]
```

Comment:
You could probably simplify this if you changed the semantics of the cache from "operation supported" to "operation not supported". Then to check if delete might succeed, just

```go
if !cache[deleteKey] {
	...
}
```

By inverting the semantic, you wouldn't need to deal with map presence detection.